yjit/src/disasm.rs

```diff
@@ -0,0 +1,212 @@
+use std::fmt::Write;
+use crate::cruby::*;
+use crate::codegen::*;
+use crate::asm::*;
+use crate::core::*;
+use crate::yjit::yjit_enabled_p;
+
+/// Primitive called in yjit.rb
+/// Produce a string representing the disassembly for an ISEQ
+#[no_mangle]
+pub extern "C" fn rb_yjit_disasm_iseq(_ec: EcPtr, _ruby_self: VALUE, iseqw: VALUE) -> VALUE {
+
+    #[cfg(not(feature = "disasm"))]
+    {
+        let _ = iseqw;
+        return Qnil;
+    }
+
+    #[cfg(feature = "disasm")]
+    {
+        // TODO:
+        //if unsafe { CLASS_OF(iseqw) != rb_cISeq } {
+        //    return Qnil;
+        //}
+
+        if !yjit_enabled_p() {
+            return Qnil;
+        }
+
+        // Get the iseq pointer from the wrapper
+        let iseq = unsafe { rb_iseqw_to_iseq(iseqw) };
+
+        let out_string = disasm_iseq(iseq);
+
+        return rust_str_to_ruby(&out_string);
+    }
+}
+
+#[cfg(feature = "disasm")]
+fn disasm_iseq(iseq: IseqPtr) -> String {
+    let mut out = String::from("");
+
+    // Get a list of block versions generated for this iseq
+    let mut block_list = get_iseq_block_list(iseq);
+
+    // Get a list of codeblocks relevant to this iseq
+    let global_cb = CodegenGlobals::get_inline_cb();
+
+    // Sort the blocks by increasing start addresses
+    block_list.sort_by(|a, b| {
+        use std::cmp::Ordering;
+
+        // Get the start addresses for each block
+        let addr_a = a.borrow().get_start_addr().unwrap().raw_ptr();
+        let addr_b = b.borrow().get_start_addr().unwrap().raw_ptr();
+
+        if addr_a < addr_b {
+            Ordering::Less
+        }
+        else if addr_a == addr_b {
+            Ordering::Equal
+        }
+        else {
+            Ordering::Greater
+        }
+    });
+
+    // Compute total code size in bytes for all blocks in the function
+    let mut total_code_size = 0;
+    for blockref in &block_list {
+        total_code_size += blockref.borrow().code_size();
+    }
+
+    // Initialize capstone
+    extern crate capstone;
+    use capstone::prelude::*;
+    let cs = Capstone::new()
+        .x86()
+        .mode(arch::x86::ArchMode::Mode64)
+        .syntax(arch::x86::ArchSyntax::Intel)
+        .build()
+        .unwrap();
+
+    out.push_str(&format!("NUM BLOCK VERSIONS: {}\n", block_list.len()));
```

Comment:
As `format!` allocates, it might be better to use [write!](https://doc.rust-lang.org/std/macro.write.html) instead as `String` implements `std::fmt::Write` (`write!` [can't fail](https://github.com/rust-lang/rust/blob/1.47.0/library/alloc/src/string.rs#L2414-L2427), so `unwrap` is safe):

```suggestion
    write!(&mut out, "NUM BLOCK VERSIONS: {}\n", block_list.len()).unwrap();
```

Reply:
Today I learned! I'm going to keep the current code though since we don't need disassembly to be super fast and `format!` shows up in intro books so it's more beginner friendly.

Reply:
I don't think it's good to keep current code -- Clippy will start to warn about this (https://rust-lang.github.io/rust-clippy/master/#format_push_string) since 1.61.0.